### ISSUE TITLE
Have a way of ensuring redux-persist has loaded data before we make any server calls

### ISSUE THREAD

**Issue by drager** [apollographql/apollo-client#604]
When using `redux-action-buffer` to wait for `redux-persist` to load stored data from local storage, the NetworkInterface middleware runs when the `APOLLO_QUERY_INIT` is fired by the component and not when it's actually executed on the reducer.

![screenshot from 2016-08-31 10-38-51](https://cloud.githubusercontent.com/assets/1957563/18122003/a4838412-6f67-11e6-9ac0-781290f671e7.png)

**Comment by drager** [comment]
@stubailo: Exactly, I have authorization using JWT on the server and I store the token in localStorage and I want to get that token and set in the middleware:

```ts
networkInterface.use([{
  applyMiddleware(req, next) {
    if (!req.options.headers) {
      req.options.headers = {}
    }
    const state = store && store.getState() || {}
    console.log('middleware', state)
    req.options.headers['Authorization'] = state.user && state.user.token
    next()
  }
}])
```

**Comment by stubailo** [comment]
So Apollo Client is more like an action creator (think redux-thunk) than anything else. The network fetch is triggered independently from the Redux actions - `APOLLO_QUERY_INIT` is not the thing that triggers the fetch at all.

So mostly this is something like "have a way to stop the Apollo action creator from firing anything until I say I'm ready".

Is there a solution that will work in the general case? How can we determine if the app is ready to start sending GraphQL requests, if there are async actions that need to be taken to become ready?

**Comment by drager** [comment]
I see... I can't say right now if there's any good general solution for this at the moment. I have to come back regarding that.

By the way. I'm using Angular 2 and the [angular2-apollo](https://github.com/apollostack/angular2-apollo) package. The desired behaviour worked just fine using Angular 2 RC-4 with apollo-client 0.4.13 and angular2-apollo 0.4.2. Then I updated to Angular 2 RC-5 and angular2-apollo to 0.4.4 and it broke. Might just have been luck that it worked before...

**Comment by stubailo** [comment]
Yeah I think you probably just got lucky that in one case the component rendered after the persisted data came back (works) and in the other the component rendered first and fired the request before the cache was loaded.

Let's keep this open and think about it a bit, I'm sure there is something. I imagine there are a lot of "startup" things people might want to do before the first query is sent.

**Comment by drager** [comment]
@stubailo: You said that: "The network fetch is triggered independently from the Redux actions", isn't this the problem? Shouldn't the network fetch be triggered when the action is triggered?

**Comment by stubailo** [comment]
That's how things work in Redux-saga, but not other action creators. Even if we did do it that way though, I don't think it would fix this issue.

**Comment by tmeasday** [comment]
@drager I've been looking at this and related issues (https://github.com/apollostack/apollo-client/issues/701#issuecomment-249084554).

@stubailo I think if we did it the "saga" way, then our action to start the saga would be deferred until the rehydration is complete, and when the saga started, the store would be populated and no network events would occur.

However, I don't know if we should change the structure of Apollo just for this use case.[1] 

@drager I think the easiest solution is really to just not render the actual page until the store has hydrated (as in [this example](https://github.com/rt2zz/redux-persist/blob/master/docs/recipes.md#delay-render-until-rehydration-complete)).

Does that make sense to you? Would it be helpful to document this somewhere?

[1] I do wonder if we should trigger network effects via thunks however (so mounting a `graphql` container would trigger a thunk, which would be able to be (in theory) manipulated by other middleware). 

However, in my playing with this stuff, it doesn't seem at all [simple](https://github.com/apollostack/apollo-client/issues/701#issuecomment-249515079) for dev tools to mess with thunks anyway. It certainly is the case that redux-persist doesn't stop the middleware from executing, so if we triggered network requests in middlewares, we'd have this problem anyway.

**Comment by Pajn** [comment]
>I think the easiest solution is really to just not render the actual page until the store has hydrated (as in this example).

The problem here is that the user has to stare at a white page for even longer, if just the requests where delayed the user would at least get to see the app shell until data has come in.
This could of course be avoided by conditionally rendering every component that does data fetching,
but that has two different problems depending on your architecture.
If you use a few "page/smart/whatever" components that does the fetching and the rest is simple functions, then conditionally rendering them would pretty much equal conditionally rendering the whole app.
If you instead does the fetching in many simpler components, then you need a lot of boilerplate and it's easy to miss the conditional render for a simple component which is only noticed if you reload the page on the particular url that component shows up.

>However, in my playing with this stuff, it doesn't seem at all simple for dev tools to mess with thunks anyway. It certainly is the case that redux-action-buffer doesn't stop the thunk middleware from executing, so if we triggered network requests in a thunk, we'd have this problem anyway. Perhaps someone has written a redux-thunk-buffer too?

That should depend on the order you specify your middlewares. If redux-action-buffer sits before redux-thunk it should buffer the execution of the thunks themselves.

**Comment by tmeasday** [comment]
> The problem here is that the user has to stare at a white page for even longer, if just the requests where delayed the user would at least get to see the app shell until data has come in.

I mean we are talking about the time it takes `redux-persist` to read localstorage and fire off an rehydrate event, no? I don't think it'd be user visible.

> If redux-action-buffer sits before redux-thunk it should buffer the execution of the thunks themselves.

Oh, you are right, this does work. (Although it's exactly what they [say not to do](https://github.com/rt2zz/redux-action-buffer#redux-persist-example), I wonder if other tools will assume that thunks will not get delayed?).

**Comment by Pajn** [comment]
> I mean we are talking about the time it takes redux-persist to read localstorage and fire off an rehydrate event, no? I don't think it'd be user visible.

Of course it probably isn't the reading from local storage it self that takes time, but the multiple required flushes of the event queue with a fully loaded main thread. If the browser were idle then it would probably not be noticeable, but when firing up a framework like React or Angular I have seen a noticeable slowdown. 
Of course the difference is much less than it would be to pre/server render the initial page, but that is not always an easy thing to do.

>Oh, you are right, this does work. (Although it's exactly what they say not to do, I wonder if other tools will assume that thunks will not get delayed?).

Hmm, I see that now. I wounder why that is though...

Depending on a not recommended feature in redux-action-buffer is of course not a good solution for Apollo. Currently I don't see any good solution to this other than avoid autorehydration from redux-persist and instead manually, and synchronously rehydrate before the store is loaded so that the initialState prameter of createStore can be used.

**Comment by tmeasday** [comment]
> Depending on a not recommended feature in redux-action-buffer is of course not a good solution for Apollo.

Probably not, but I think it does add to the case that Apollo should dispatch network events via thunks. Will consider this some more.

**Comment by helfer** [comment]
Removing this from the milestone, since we didn't touch this during the refactor. I think I now have a better idea what this is about, so we can take a fresh look at it after 0.5

**Comment by carobeta** [comment]
@drager  I use redux-persist too, but my apollo state never rehydrated. I need to implement something for that?  Thanks!

**Comment by peitalin** [comment]
@carobeta 
Not sure if this helps, but you can inject the redux-persist state directly into apollo-client, with
`getStoredState({ storage: localStorage }, (err, rehydratedState) => { ... }`

e.g:
```
getStoredState({ storage: localStorage }, (err, rehydratedState) => {

	const client = new ApolloClient({
		networkInterface: networkInterface,
		initialState: { apollo: { data: rehydratedState.apollo.data }},
	});

	let reduxStore = createStore(
		combineReducers({
			...other_reducers,
			apollo: client.reducer()
		}),
		rehydratedState,
		compose(
			// applyMiddleware(client.middleware()), // don't use this
			// rehydrate(), // don't use this
		)
	);

	const persistor = createPersistor(reduxStore, { storage: localStorage })
})

```

**Comment by magbicaleman** [comment]
@peitalin how do you export the client out ?

**Comment by peitalin** [comment]
I had it in a react component so just assigned it to this.client.

But I don't think this is a good approach. It wasn't reliably persisting state (I suspect it was a fluke that it worked then)

**Comment by magbicaleman** [comment]
I wish there was a different approach, but this is the one I found here that would delay your whole app. [Delay Render Until Rehydration Complete](https://github.com/rt2zz/redux-persist/blob/master/docs/recipes.md#delay-render-until-rehydration-complete)

**Comment by mvanlonden** [comment]
Any updates here? Is there a recognized best approach?

**Comment by mvanlonden** [comment]
Nevermind, this has been fixed by the recent update `1.9` https://github.com/apollographql/apollo-client/issues/424#issuecomment-316529011
